requests/utils.py

```diff
@@ -579,12 +579,19 @@ def select_proxy(url, proxies):
     """
     proxies = proxies or {}
     urlparts = urlparse(url)
-    if urlparts.hostname is None:
-        proxy = None
-    else:
-        proxy = proxies.get(urlparts.scheme+'://'+urlparts.hostname)
-    if proxy is None:
-        proxy = proxies.get(urlparts.scheme)
+    proxy = None
+    if urlparts.hostname is not None:
+        proxy_keys = [
+            'all://' + urlparts.hostname,
+            'all',
+            urlparts.scheme + '://' + urlparts.hostname,
+            urlparts.scheme,
+        ]
+        for proxy_key in proxy_keys:
+            if proxy_key in proxies:
+                proxy = proxies[proxy_key]
+                break
+
```

Comment:
This code has logically changed. Before this PR, if the hostpart of the url was `None`, we'd lookup the proxy by scheme alone. Now we don't do that: if it's `None`, we return no proxy at all because we never enter the `if` block. Care to refactor this?